Try another approach to fixing the sibling inserter in Firefox

jasmussen opened the pull request:
This PR partially fixes #7508. It is an alternate version of #7525, though much the same.

In #7220 we introduced a new behavior for the sibling inserter, which requires a click on the plus in the center as opposed to just clicking between two blocks.

Turns out it was sort of a race condition between onClick and onMouseDown, the latter which fires first. So in a way, the Firefox and Safari behavior of selecting the block (which is selected onMouseDown) as opposed to clicking the sibling inserter was the correct one.

This PR "fixes" it by also making the sibling inserter use onMouseDown. But in addition to this, it uses onClick as well, so it's still keyboardable.

The net effect is that both work, with the added benefit that in Firefox and Safari, the block that you're hovering isn't briefly "selected" when you're clicking.

GIF:

![sibling](https://user-images.githubusercontent.com/1204802/41849815-a14cafbe-7882-11e8-87ef-e3f85e66b1d0.gif)

Note, though, that when clicking, text focus isn't set on the newly created sibling block. This is something we'll want to fix before this can be merged.

jasmussen committed:
Try another approach to fixing the sibling inserter in Firefox

This PR partially fixes #7508. It is an alternate version of #7525, though much the same.

In #7220 we introduced a new behavior for the sibling inserter, which requires a click on the plus in the center as opposed to just clicking between two blocks.

Turns out it was sort of a race condition between onClick and onMouseDown, the latter which fires first. So in a way, the Firefox and Safari behavior of selecting the block (which is selected onMouseDown) as opposed to clicking the sibling inserter was the correct one.

This PR "fixes" it by also making the sibling inserter use onMouseDown. But in addition to this, it uses onClick as well, so it's still keyboardable.

The net effect is that both work, with the added benefit that in Firefox and Safari, the block that you're hovering isn't briefly "selected" when you're clicking.

tofumatt reviewed (dismissed):
This could be DRY'd up, but since I'll take a shot at getting the focus to work I'll just do that ^_^

tofumatt committed:
Use the onClick handler for mouseDown

youknowriad reviewed (approved):
LGTM 👍 

Maybe we can add a comment clarifying why we need this duplication. Also, do you have any idea if we can replace the parent `onMouseDown` event with `onClick`? This could be more impactful but I wonder why we're using `onMouseDown` in the first place.

jasmussen committed:
Fix issues in Firefox and Safari

Turns out this issue has been present in some hidden form for a long time, and _partially fixed_ in Chrome. But there has always been a race condition, it looks like, between onClick, onFocus and onMouseDown.

This commit simply adds the "onFocusInserter" action to the onMouseDown event as well, which seems to solve the issue.

You still briefly see the block being selected and the toolbar appearing in Firefox and Safari, but at least this can work as a hotfix pending further investigation or improvements.

youknowriad reviewed (changes requested):
Just noticed the breakage in the e2e tests, we might want to take a look at those ;)

jasmussen commented:
Thanks for the fast review, and will look at tests in a second. Just wanted to clarify that I just pushed a change in https://github.com/WordPress/gutenberg/pull/7530/commits/743c21995c96d482f41f65ce837183e4d85fd755 which changes it so the onFocusInserter is attached to onMouseDown as well, and this fixes the issue for Firefox and Safari, without us losing focus. Behold:

![safari](https://user-images.githubusercontent.com/1204802/41897805-a0e0f6ec-7928-11e8-9ae0-f635bdb21a0e.gif)

The issue where the block toolbar briefly appears is still there, but this is an improvement in that it now works as intended.

jasmussen committed:
Add comment.

jasmussen commented:
Okay I think the latest two commits might have fixed the e2e tests. Final review?

aduth commented:
I don't think the changes to `package-lock.json` were intended here? I see local changes anytime I run `npm install`.

aduth commented:
See #7641

aduth commented:
This is purely a styling issue: #7642
